### PR TITLE
[FLINK-20306][table-planner-blink] Throw a meaningful exception when accessing temporal table as of constant timestamp

### PR DESCRIPTION
**Pull request by wuchong**
## What is the purpose of the change

Currently, the planner throws a cryptic exception message when accessing temporal table with constant timestamp.

```
SELECT * from RatesHistory FOR SYSTEM_TIME AS OF TIMESTAMP '2020-11-11 13:12:13';

org.apache.calcite.plan.RelOptPlanner$CannotPlanException: There are not enough rules to produce a node with desired properties: convention=STREAM_PHYSICAL, FlinkRelDistributionTraitDef=any, MiniBatchIntervalTraitDef=None: 0, ModifyKindSetTraitDef=[NONE], UpdateKindTraitDef=[NONE].
Missing conversion is FlinkLogicalSnapshot[convention: LOGICAL -> STREAM_PHYSICAL]
There is 1 empty subset: rel#987:RelSubset#44.STREAM_PHYSICAL.any.None: 0.[NONE].[NONE], the relevant part of the original plan is as follows
977:FlinkLogicalSnapshot(period=[2020-11-11 13:12:13])
  975:FlinkLogicalCalc(subset=[rel#976:RelSubset#43.LOGICAL.any.None: 0.[NONE].[NONE]], select=[CAST(Reinterpret(CAST(timestamp))) AS currency_time, currency, CAST(rate) AS rate])
    962:FlinkLogicalTableSourceScan(subset=[rel#974:RelSubset#42.LOGICAL.any.None: 0.[NONE].[NONE]], table=[[default_catalog, default_database, RatesHistory, watermark=[CAST($2):TIMESTAMP(3)]]], fields=[currency, rate, timestamp])
```

We can throw a better exception to indicate this is not supported yet. 

## Brief change log

- throw a better exception when constructing `FlinkLogicalSnapshot` via `isValid(...)` interface.

## Verifying this change

- Added validation tests which can reproduce the problem.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Review by dawidwys — APPROVED** [review]
LGTM, once azure is green

**Review by dawidwys — CHANGES_REQUESTED** [review]
Actually there are related test failures, which should be fixed first.

**Comment by wuchong** [comment]
The failed case is related to FLINK-20328.

This is passed in my private account: https://dev.azure.com/imjark/Flink/_build/results?buildId=356&view=results

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/common/CommonTemporalTableJoinRule.scala [hunk 1]
```diff
@@ -44,20 +44,15 @@ trait CommonTemporalTableJoinRule {
...
+          "'FOR SYSTEM_TIME AS OF' left table's time attribute field.")
     }
 
-    snapshot.getPeriod.getType match {
-      // supports both event-time and processing time
-      case t: TimeIndicatorRelDataType =>
-      case _ =>
-        throw new TableException("Temporal table join currently only supports " +
-          "'FOR SYSTEM_TIME AS OF' left table's time attribute field")
-    }
```
Comment (godfreyhe) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/common/CommonTemporalTableJoinRule.scala:60]: Do we check the period type in other place ?

Comment (wuchong) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/plan/rules/common/CommonTemporalTableJoinRule.scala:60]: I merged it into the above check.

**Review by godfreyhe — APPROVED** [review]

**Comment by wuchong** [comment]
Merging...

**Pull request merged**